### PR TITLE
Use wp_resource_hints to add urls to dns-prefetch directives.

### PR DESCRIPTION
**Pull request by asvinb**
## Summary

Addresses issue #2203 

## Checklist

- [x] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 4.7 and PHP 5.6.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [x] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

### CODE REVIEW

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Thanks, @asvinb. It looks okay, but we still need to add DNS prefetch for the Analytics module tag.

Plus, I think we can extract the common functionality into a trait class to reduce duplicate code.  Let's create a new `includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php` with the following trait:

```php
trait Tag_With_DNS_Prefetch_Trait {
    protected function get_dns_prefetch_hints_callback( $url ) {
        return function( $urls, $relation_type ) use ( $url ) {
            if ( 'dns-prefetch' === $relation_type ) {
                $urls[] = $url;
            }
            return $urls;
        };
    }
}
```

Then add this trait to AdSense, Analytics and TagManager tags and register `wp_resource_hints` filters:

```php
add_filter(
    'wp_resource_hints',
    $this->get_dns_prefetch_hints_callback( '//www.googletagmanager.com' ),
    10,
    2
);
```

**Comment by asvinb** [comment]
> Thanks, @asvinb. It looks okay, but we still need to add DNS prefetch for the Analytics module tag.
> 
> Plus, I think we can extract the common functionality into a trait class to reduce duplicate code. Let's create a new `includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php` with the following trait:
> 
> ```
> trait Tag_With_DNS_Prefetch_Trait {
>     protected function get_dns_prefetch_hints_callback( $url ) {
>         return function( $urls, $relation_type ) use ( $url ) {
>             if ( 'dns-prefetch' === $relation_type ) {
>                 $urls[] = $url;
>             }
>             return $urls;
>         };
>     }
> }
> ```
> 
> Then add this trait to AdSense, Analytics and TagManager tags and register `wp_resource_hints` filters:
> 
> ```
> add_filter(
>     'wp_resource_hints',
>     $this->get_dns_prefetch_hints_callback( '//www.googletagmanager.com' ),
>     10,
>     2
> );
> ```

Nice! I didn't know about that. PR updated, can you check?

File: includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Trait for adding the dns-prefetch directive to a url.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+trait Tag_With_DNS_Prefetch_Trait {
+	/**
```
Comment (eugene-manuilov) [inline includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php:21]:
```suggestion
trait Tag_With_DNS_Prefetch_Trait {

	/**
```

File: includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+	 */
+	protected function get_dns_prefetch_hints_callback( $url ) {
+		return function( $urls, $relation_type ) use ( $url ) {
+			if ( 'dns-prefetch' === $relation_type ) {
+				$urls[] = $url;
+			}
+			return $urls;
+		};
+	}
+}
```
Comment (eugene-manuilov) [inline includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php:37]:
```suggestion
	}

}
```

File: includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+/**
+ * Trait for adding the dns-prefetch directive to a url.
+ *
+ * @since n.e.x.t
+ * @access private
+ * @ignore
+ */
+trait Tag_With_DNS_Prefetch_Trait {
+	/**
+	 * Adds the dns-prefetch directive to a specified URL.
```
Comment (eugene-manuilov) [inline includes/Core/Tags/Tag_With_DNS_Prefetch_Trait.php:22]:
```suggestion
	 * Gets a callback that can be used for the wp_resource_hints filter to set the dns-prefetch directive for a specified URL.
```

**Review by eugene-manuilov — CHANGES_REQUESTED** [review]
Looks good to me. Thanks, @asvinb. There are a few formatting issues that I'll apply and then I'll approve this PR.

**Review by eugene-manuilov — APPROVED** [review]

**Pull request merged**